File: lua/sort.lua
--[[
Sorting package for implementing simple sort algorithms. This depends on a 
data structure that has 4 methods (inspired by Go's Sort interface):
 
* len() - Length of elements
* less(i, j) - Given two indexes is the first less than the second
* swap(i, j) - Swap the elements locations

--]]
local sort = {} -- Sort package

local function partition(sortable, low, high)
	local pivot = high
	for j = low, high do
		if sortable:less(j, pivot) then
			sortable:swap(j, low)
			low = low + 1
		end
	end
	sortable:swap(low, high)
	return low
end
-- Logic behind most q sorts
local function qsort(sortable, low, high)
	if low > high then
		return
	end
	local p = partition(sortable, low, high)
	qsort(sortable, low, p - 1)
	qsort(sortable, p + 1, high)
end

function sort.quick(sortable)
	qsort(sortable, 1, sortable:len())
end
return sort

File: zig/sort.zig
const std = @import("std");

fn swap(comptime T: type, elements: []T, i: usize, j: usize) void {
    const x = elements[i];
    const y = elements[j];
    elements[i] = y;
    elements[j] = x;
}
fn partition(comptime T: type, elements: []T, low: usize, high: usize) usize {
    const pivot = high;
    var i = low;
    var j = low;
    while (j < high) : (j += 1) {
        if (elements[j] < elements[pivot]) {
            swap(T, elements, j, i);
            i += 1;
        }
    }
    swap(T, elements, i, high);
    return i;
}
fn qsort(comptime T: type, elements: []T, low: usize, high: usize) void {
    if (low > high) {
        return;
    }
    const p = partition(T, elements, low, high);
    if (p < 1) {
        return;
    }
    qsort(T, elements, low, p - 1);
    qsort(T, elements, p + 1, high);
}

pub fn quick(comptime T: type, elements: []T) void {
    qsort(T, elements, 0, elements.len - 1);
}

const expect = std.testing.expect;
test "quick sort" {
    var testInts = [_]u32{ 3, 1, 3, 5, 1, 1, 2, 6 };
    quick(u32, &testInts);
    try expect(testInts[0] == 1);
    try expect(testInts[6] == 5);
    try expect(testInts[7] == 6);
    for (testInts) |num| {
        std.debug.print("Value: {d}\n", .{num});
    }
}
